Support Math keyword in native_functions.yaml.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44993 Define macro for DispatchKey::toString.
* #44933 Align casing in test_dispatch with dispatch keys.
* **#44556 Support Math keyword in native_functions.yaml.**
* #44354 Add alias dispatch key Math.

Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

Commit:
Support Math keyword in native_functions.yaml.

[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."




[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/README.md:
```diff
@@ -277,6 +277,18 @@ them the same thing!)
+Available backend options can be found at
+https://github.com/pytorch/pytorch/blob/master/tools/codegen/gen.py#L970.
+In addition to backends above, we also support keyword `Math` which is an alias
+that maps to all backends. In other words, function registered to `Math` key
+should be a plain mathematical composition of other `at::` functions and works for any backend.
```
(drive-by comment): Are things registered to the Math key required to be functional (e.g., no in-place operators), or are in-place operations OK?

ailzhang replied:
It doesn't have to be functional. I'd think of `Math` key being used to store the old `composite` implementations. 
Imagine we used to implement `native_group_norm` using `at::native_batch_norm` (it doesn't need to be functional), once we add an optimized kernel for certain backend, say
 ```
func: native_group_norm
CPU: group_norm_cpu
CUDA: group_norm_cuda
# Math: old_native_group_norm
```
`Math` is a way to tell all other backends you can still use `old_native_group_norm` without issues. Before this PR, once we add `CPU: group_norm_cpu`, backends like XLA are suddenly required to register a kernel to `native_group_norm` otherwise there's no kernel available for that key. 
Hope this explains it a little better, let me know if you have followup questions!

zou3519 commented on aten/src/ATen/native/README.md:
```diff
@@ -277,6 +277,18 @@ them the same thing!)
 If two backends have the same dispatch function, you can write `CPU, CUDA: func`
 to reuse the same function name in both cases.
 
+TODO: expand this section.
```
(another drive-by comment): I don't know what the Math key is but it looks interesting. Would love to see an explanation of what it's used for here

ailzhang replied:
https://fb.quip.com/jiYeAa58PDX0 explains it pretty well. But yea I should improve docs here as well. ;)

ezyang replied:
Yeah, let's publish the design on a GitHub issue so non-FB people can see it.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -370,7 +370,8 @@
 REGISTRATION_DECLARATION = CodeTemplate("""\
-${return_type} ${api_name}(${declaration_formals}); // {"schema": "${schema_string}", "compound": "${compound}"}
+${return_type} ${api_name}(${declaration_formals}); \
+// {"schema": "${schema_string}", "compound": "${compound}", "has_math_kernel": "${has_math_kernel}"}
```
Wondering if we can just port this codegen straight to the new codegen, then we don't have to expose `has_math_kernel` in Declarations.yaml.

ailzhang replied:
Unfortunately it has to stay in autograd codegen for now since we need access to `whether an op has derivative` to cover DispatchSub ops. Whether the kernel comes from disptach stub is not available anywhere in codegen and can only be approximated by the fact that almost all dispatch stub ops have derivatives). So we need to add this information in native_functions.yaml before we could move this codegen. More details can be found at: https://fb.quip.com/1B61AfNcuozU

ezyang replied:
OK. We'll be out of this situation soon, but this is fine for this PR.

ezyang commented on tools/codegen/gen.py:
```diff
@@ -1016,16 +1017,30 @@ def make_file_manager(install_dir: str) -> FileManager:
             compute_type_method(None, target=Target.REGISTRATION, op_registration_whitelist=op_registration_whitelist),
+            native_functions)) +
+        list(mapMaybe(
+            compute_type_method('Math', target=Target.REGISTRATION, op_registration_whitelist=op_registration_whitelist),
```
This happens to work but the reason is not so good. When you dunk this in `function_registrations` it gets dumped in:

```
TORCH_LIBRARY(aten, m) {
  ${function_registrations};
```

This doesn't specify a dispatch key specifically (unlike the CPU/CUDA files). Everything works out because we always generate `torch::dispatch`:

```
                # Annotate it with dispatch information if necessary
                #
                # NB: In the ordinary, TypeDerived code generation work flow, specification
                # of the backend is handled by the enclosing block, so the torch::dispatch
                # invocation here is strictly unnecessary.  However, in the fbcode mobile
                # only workflow using per-op registration, these registrations will get dumped
                # in a TORCH_LIBRARY_FRAGMENT that does not have an ambient backend.  So
                # the torch::dispatch specification here is important!  See
                # Note [Redundancy in registration code is OK] for how we handle redundant info.
                if dispatch is not None:
                    payload = f"torch::dispatch(DispatchKey::{dispatch},\n{payload})\n"
```

With this change, this comment is no longer true, since you're also relying on the torch::dispatch call to make sure Math goes to the right place. But it seems to me that it would be clearer if you just made a separate `math_registrations` entry and then put it in a dedicated `Math` registration block, at which point the comment above would remain true.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -654,12 +655,12 @@ def gen_variable_type(out, aten_declarations, template_path):
                 REGISTRATION_DECLARATION.substitute(declaration,
                                                     declaration_formals=declaration_formals,
-                                                    compound='false'))
+                                                    compound='False'))
```
Not JSON anymore? lol

Commit:
Update on "Support Math keyword in native_functions.yaml."




[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."




[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."




[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/group_norm.cpp:
```diff
@@ -133,5 +133,29 @@ Tensor group_norm(
 DEFINE_DISPATCH(GroupNormKernel);
 DEFINE_DISPATCH(GroupNormBackwardKernel);
 
+std::tuple<at::Tensor, at::Tensor, at::Tensor> math_group_norm(
```
For CR purposes it would be helpful to cite where this is coming from

ailzhang replied:
I wrote the pytorch/xla version based on the old pytorch composite implementation. So it should be fine here ;)

ezyang commented on aten/src/ATen/test/math_kernel_test.cpp:
```diff
@@ -0,0 +1,40 @@
+
+// Ideally we want to test both forward and backward on math kernels but I
+// haven't found an easy way to do it.  Currently we only test forward here
+// and rely on backward tests of each at:: function used in math kernels.
```
This is definitely something we should work on making easier, though I'm not exactly sure how. Perhaps APIs for supporting static runtime would make it easier, as they would make it possible to ask for a specific kernel.

Absent machinery like this, I think the easiest way to do this is to just make the math kernel available at another name (i.e., make a new `native_functions.yaml` entry; and then just do the testing from Python. Maybe @mruberry has some other ideas.

mruberry replied:
Without too much context I'll also vote for (re)using Python.

ezyang commented on aten/src/ATen/native/group_norm.cpp:
```diff
@@ -133,5 +133,29 @@ Tensor group_norm(
+  }
+  at::Tensor mean = std::get<1>(outputs).view({N, group});
+  at::Tensor rstd = std::get<2>(outputs).view({N, group});
+  return std::make_tuple(out, mean, rstd);
```
I'm going to trust that this was lifted directly from XLA and has already been code reviewed.

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -277,6 +277,18 @@ them the same thing!)
+Available backend options can be found at
+https://github.com/pytorch/pytorch/blob/master/tools/codegen/gen.py#L970.
+In addition to backends above, we also support keyword `Math` which is an alias
+that maps to all backends. In other words, function registered to `Math` key
```
Do we want to note here that `Math` maps to all autograd keys as well? Or is this doc expressly written to avoid pulling in a discussion of autograd?

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -277,6 +277,18 @@ them the same thing!)
+should be a plain mathematical composition of other `at::` functions and works for any backend.
+
+If you add `dispatch` section to any API that didn't have it before, you **have to** move
+the old implementation to `Math` field so that it's still available for other backends to use.
```
We can also enforce this, right? (Maybe it's further down the PR...)

ailzhang replied:
Good point! We currently cannot enforce it easily since we don't track whether ops has dispatch sections and whether they change. But this should be possible in the terminal state where we can track this through our API provided to backend extensions. ;) So I'd save this as a followup task. ;)

bhosmer replied:
Ah yeah - actually I oversimplifying , picturing a world where we could simply insist that if you had any explicit dispatch entries at all, then `Math` had to be one of them (say, by adding a semantic check in `NativeFunction.from_yaml`). But of course we'd need a codemod (or something) before our current declarations would fail that test, so yeah - followup task sounds right!

Commit:
Update on "Support Math keyword in native_functions.yaml."




[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."




[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."




[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]

bhosmer reviewed (approved):
Looks good!

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]

Commit:
Update on "Support Math keyword in native_functions.yaml."


Differential Revision: [D23698386](https://our.internmc.facebook.com/intern/diff/D23698386)

[ghstack-poisoned]
